### PR TITLE
[fix](video) Align complete Python and native video contracts

### LINKED ISSUE #777: [bug](video) Python and native disagree on the complete video input/output contract
### Affected area

DuckDB or SQL execution

### What happened?

The video Python and native backends do not yet have the same complete input/output contract. Switching the backend can change metadata or selection semantics even when the decoded images match. This issue follows the RGB conversion fix in #774 / #776.

Observed in the differential audit on commit `73e5650e7406790946f72ed97b728b85fd682f74`:

- A Matroska H.264 fixture returns first-frame `frame_dts=125` through Python and `NULL` through native, while RGB and PTS match.
- Python metadata substitutes container duration when the selected video stream duration is unknown; native reads only the video stream duration. Container duration is not exposed separately.
- Python preserves a decoder-reported frame duration of zero; native currently emits NULL for zero.
- Python uses exact rational time selection, while native sampling adds a floating-point tolerance.
- The connection-bound `VideoFrameSource` entry point returns Tensor pixels with Python and IMAGE pixels with native. Its bound relation must have the same complete output type and fields across backends.
- Index construction, indexed reads and scan diagnostics currently reject the Python backend. This capability difference must be addressed explicitly as part of the complete contract, not silently omitted from comparisons.

Expected contract:

1. `duration` denotes only the selected video stream duration. Unknown values remain NULL. Expose the whole-file duration as `container_duration`; never substitute one for the other.
2. `frame_count` preserves known counts and NULL for unknown counts; do not infer it from duration multiplied by FPS.
3. `video_frames(NULL)` returns NULL; a valid selection with no matching frames returns an empty list.
4. The same public API, source and parameters produce the same frame selection, RGB bytes, complete temporal metadata, NULL behavior and public error behavior across Python/native. Include sequential, indexed, keyframe and streaming paths, including the relation returned by `connection.from_datasource(VideoFrameSource(...))`. The built-in source must return IMAGE pixels through either backend; standalone Python task execution retains its own declared task schema.
5. Investigate the packet-to-decoder-to-frame DTS path before deciding the fix. Keep actual provenance; do not fill DTS from PTS or erase fields to make a test pass.
6. Add an independent Python implementation of index construction, the index codec and validation, keyframe seeking, indexed selection, and scan statistics. Keep the native algorithm independent. Both backends must read each other's indexes and produce identical complete outputs and counters. SQL on the Python backend enters Python through the existing C++ execution bridge; this must not delegate indexing or decoding to the native video extension.
7. Differential results report pixels, selection, metadata, and NULL/error behavior separately while still requiring full public-output equality. Do not exclude DTS or use approximate assertions to hide differences.

### Reproduction

Use a non-editable Vane installation and a matching native video artifact, with PyAV, NumPy and Pillow installed. Set `VANE_TEST_NATIVE_VIDEO_EXTENSION` to that artifact, then run:

```python
import os
from pathlib import Path
from tempfile import TemporaryDirectory

import av
import numpy as np
import vane

with TemporaryDirectory() as directory:
    path = Path(directory) / "video.mkv"
    with av.open(str(path), "w") as output:
        stream = output.add_stream("libx264", rate=8)
        stream.width, stream.height, stream.pix_fmt = 64, 48, "yuv420p"
        stream.codec_context.gop_size = 8
        stream.codec_context.max_b_frames = 2
        stream.options = {"crf": "18", "sc_threshold": "0", "threads": "1"}
        y, x = np.indices((48, 64))
        for i in range(24):
            pixels = np.stack(((x * 7 + i * 19) % 256,
                               (y * 11 + i * 31) % 256,
                               ((x ^ y) * 13 + i * 47) % 256), axis=-1).astype("uint8")
            frame = av.VideoFrame.from_ndarray(pixels, format="rgb24")
            for packet in stream.encode(frame):
                output.mux(packet)
        for packet in stream.encode():
            output.mux(packet)

    with vane.connect(config={"video_backend": "python"}) as python, vane.connect(
        config={"allow_unsigned_extensions": "true"}
    ) as native:
        native.load_extension(os.environ["VANE_TEST_NATIVE_VIDEO_EXTENSION"])
        native.execute("SET video_backend='native'")
        for name, connection in [("python", python), ("native", native)]:
            print(name, connection.execute(
                "SELECT video_metadata(video_file($1))", [str(path)]
            ).fetchone()[0])
            first = connection.execute(
                "SELECT video_frames(video_file($1))", [str(path)]
            ).fetchone()[0][0]
            print(name, {key: first[key] for key in
                         ("frame_index", "frame_pts", "frame_dts", "frame_duration")})
```

The native artifact must match the installed engine SourceID. Unsigned loading above is only for a locally built test artifact.

### Environment

- OS / architecture: Linux x86-64
- Python: 3.12.3
- PyAV: 17.1.0
- Native FFmpeg: 8.1.1
- Audited Vane: `73e5650e7406790946f72ed97b728b85fd682f74`
- Follow-up implementation rebased onto upstream/main `a4cd645797` after #776 merged; initial implementation commit `b5137f8926`
- Local SQL execution; Ray coverage is required during final validation

### Additional context

Acceptance cases include video shorter than audio, unknown stream duration/count, VFR, offset timestamps, sub-frame sampling boundaries, MP4/Matroska with B frames, NULL inputs, malformed inputs, resource limits, sequential/indexed reads and keyframe/streaming selection. Full output records and public validation/error behavior are compared without removing differing metadata.

The requested workflow is issue first, fix/review until two consecutive local reviews are clean, then one consolidated build/test phase, followed by a template-based PR and two consecutive manual GitHub `@codex review` passes. No compatibility modes or fallback paths are requested. The three pre-existing release-test failures are tracked separately in #775.

### PR DESCRIPTION
**Pull request by kaka11chen**
## Summary

Switching video backends could change DTS, metadata and sampling, and the Python backend rejected explicit indexes. Match probe and decoder initialization, preserve stream/container duration separately and decoded zero duration, and use exact rational native time selection. Complete record comparisons include every timestamp and every RGB byte.

Implement index construction, parsing, verified FILE reads, seeking, selection and scan statistics independently in Python/PyAV and native/FFmpeg. SQL dispatch enters Python through the base C++ execution bridge without invoking the native video extension. Both index implementations use the same version 2 representation and preserve original sequential DTS/duration after a verified seek. Both parsers validate the construction read counter against the full-source hash pass and global decoder read bound. Connection-bound `VideoFrameSource` now returns generic IMAGE with both backends and enforces the same hard payload and parameter bounds. Fetched engine IMAGE values follow the upstream UInt8 HWC ndarray contract, with exact recursive array/record assertions. Preserve video exception categories through the Python Arrow scan boundary, including admission cancellation, and apply the combined FILE/index metadata budget before backend dispatch.

## Related issue

close: #777

Rebased onto upstream/main `04e68099e6`, including #776 and the UInt8 IMAGE foundation in #773. The unrelated release failures are tracked in #775.

## Documentation impact

- [ ] No user-facing documentation update is required.
- [x] Documentation is updated in this PR or in a linked website PR.
- [ ]  A follow-up issue is required in
      `AstroVela/vane-website`.

`VIDEO_FRAME_API.md` defines duration, NULL, time selection, complete provenance, index layout and streaming error behavior. Metadata adds `container_duration`; unknown stream duration stays NULL. Version 1 indexes require rebuilding. No compatibility modes or fallback paths are introduced.

## Validation

Two consecutive clean local reviews of the rebase integration preceded the consolidated Release build/test phase on `4e03545116b26019e73eb5e23f185ac91980ab28`. Root pre-commit, including Ruff and mypy, workspace formatting and `git diff --check` passed.

Two consecutive manual GitHub `@codex review` rounds on unchanged rebased commit `4e03545116b26019e73eb5e23f185ac91980ab28` reported no major issues and no new inline findings: [round 1](https://github.com/AstroVela/vane/pull/778#issuecomment-5573336180), [round 2](https://github.com/AstroVela/vane/pull/778#issuecomment-5573422252).

- Non-editable incremental Release build with matching image/video artifacts; test-only signed provider wheels for Ray. Linux x86-64, Python 3.12.3, PyAV 17.1.0, FFmpeg 8.1.1. The integration-test signing key is enabled only in this local build.

```bash
SKBUILD_BUILD_DIR="$PWD/build/python-release" SKBUILD_CMAKE_BUILD_TYPE=Release \
CMAKE_BUILD_PARALLEL_LEVEL=12 uv pip install --python .venv/bin/python . \
  --no-build-isolation --no-deps \
  '-Ccmake.define.VANE_LOADABLE_EXTENSIONS=image;video' \
  '-Ccmake.define.VANE_ENABLE_TEST_EXTENSION_SIGNING_KEY=ON'
cmake --build build/python-release --target vane_loadable_extensions --parallel 12
```

- `scripts/run_installed_pytest.sh tests/fast/test_video_file.py tests/fast/test_video_reader.py tests/fast/test_video_optional_deps.py tests/fast/test_file_reader.py tests/fast/test_datasource_distributed_scan.py tests/fast/test_udf_process.py -m 'not real_ray' --tb=short -q`: **317 passed**.
- `scripts/run_installed_pytest.sh tests/fast/test_video_backend_contract.py tests/fast/test_video_expressions.py tests/fast/test_read_video_frames.py tests/fast/test_video_index.py -m 'not real_ray' --tb=short -q`: **306 passed**. Includes rechecksummed invalid construction-byte counters, accepted bounds and rejection before FILE I/O through inspection, scalar null mode and streaming skip mode.
- IMAGE storage/materialization integration: `scripts/run_installed_pytest.sh tests/fast/test_image.py tests/fast/test_fixed_shape_image.py tests/fast/test_image_file.py -m 'not real_ray' --tb=short -q`: **245 passed, 1 deselected**.
- Native media integration: `scripts/run_installed_pytest.sh tests/fast/test_native_media_extensions.py tests/fast/test_native_media_validation.py -m 'not real_ray' -k 'not audio' --tb=short -q`: **38 passed**. Artifact paths supplied using `VANE_TEST_NATIVE_VIDEO_EXTENSION` and `VANE_TEST_NATIVE_IMAGE_EXTENSION`.
- Repeated differential audit of the original seven synthetic clips: all **99** jointly successful Python/native cases now match every recorded timestamp and RGB byte; **7** shared out-of-range errors counted separately. Sequential/indexed and cross-backend indexed outputs match. A before/after comparison also preserves every normalized successful frame record and error category from the prior reviewed head. The committed matrix additionally checks identical index bytes, cross-reading, full FILE provenance and exact scan counters on seven fixture families, including AVI, B frames, offset timestamps, VFR, NTSC and audio extending beyond video.

- With `VANE_TEST_NATIVE_MEDIA_PROVIDERS=1`, `scripts/run_installed_pytest.sh tests/fast/test_video_reader.py tests/fast/test_ray_video_expressions.py tests/fast/test_ray_read_video_frames.py tests/fast/test_ray_video_index.py tests/fast/test_ray_native_media_extensions.py -m real_ray -k video --tb=short -q`: **21 passed, no skips**, using signed providers with unsigned loading disabled. Python/native index builds, explicit index scans, scalar outputs and partitioned streaming pass on Ray. In total, **927 unique affected tests pass**.

- `scripts/run_release_tests.sh`: **1020 passed, 3 failed** in the non-Ray shard. These are the same baseline failures tracked in #775: `test_ray_worker_manager_drop_fans_out_after_result_payload_release_failure`, `test_wait_fte_query_propagates_status_errors`, and `test_wait_fte_query_release_failure_preserves_failed_handle_and_releases_rest`. They expect injected error text but receive `error detail exceeds 4096 bytes and was omitted`. The release gate is **not green** in this environment.
- The release launcher stops after that first shard. The same 15-file release list is therefore run separately with `-m 'not external_service and real_ray'`: **18 passed, 1032 deselected**.

## Checklist

- [x] The change is focused and includes tests or a reason tests are unnecessary.
- [x] Public behavior and compatibility impact are documented.
- [x] New dependencies, copied code, model assets, and datasets have compatible
      licenses and are recorded where required.
- [x] No credentials, private endpoints, personal paths, generated data, model
      weights, or build artifacts are included.
- [x] Security implications of UDFs, serialization, remote code, network access,
      and untrusted input have been considered.
- [x] Native changes were compiled; Python-only, documentation, and workflow
      changes passed relevant checks.

The video feature now directly declares Boost.Multiprecision, already covered in the repository's transitive native license inventory. Test clips are generated locally. Index checksums detect corruption, not malicious replacement; all FILE access remains governed by the executing connection and no codec receives an ambient source URL.

### CODE REVIEW

**Comment by kaka11chen** [comment]
@codex review

Please review the complete PR at `d09c74950a67532a3194da861a92ffc686069890`. In particular, check independent Python/native index algorithms, exact complete output and timestamp parity, governed FILE I/O and index validation, exception preservation, and connection-bound IMAGE scans. No compatibility modes or fallback paths are required.

File: vane/_video_index.py [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+        raise vane.InvalidInputException("invalid video index format")
+    with memoryview(data) as view:
+        if _digest(view[:-32], check) != data[-32:]:
+            raise vane.InvalidInputException("video index integrity check failed")
+    if struct.unpack_from("<4Q", data, 8) != _versions(av):
+        raise vane.InvalidInputException("video index requires the codec build that created it")
+    source_size = struct.unpack_from("<Q", data, 40)[0]
+    if data[48:80] != _digest(vane.__git_revision__.encode(), check):
+        raise vane.InvalidInputException("video index requires the engine SourceID that created it")
+    numerator, denominator, origin, build_bytes, blocks, frames = struct.unpack_from("<QQqQQQ", data, 112)
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply vane/_video_index.py:128]: Fixed in 9410f8dcfd. Both independent parsers now require source_size <= build_bytes <= source_size + 4 * 16 GiB before opening a FILE. The full-source hash pass is separate from the decoder's bounded physical reads; the per-build input limit is not serialized, so this uses the valid global bound.

The regression cases recompute the checksum after changing the count to 0, source_size - 1, the maximum + 1, and UINT64_MAX, and check inspection, scalar null mode and streaming skip mode. Both accepted boundary values are covered as well.

Validation: two clean local review rounds, non-editable Release/native artifact rebuild, and 657 non-Ray tests passed on this fix (317 Python/FILE/DataSource/UDF, 302 complete video/index contract cases, 38 native media cases). The signed-provider Ray rerun also passed all 21 cases, for 678 affected tests passing. Release rechecks: 1020 non-Ray tests passed with the same three existing #775 failures, and all 18 real-Ray release cases passed.

**Comment by kaka11chen** [comment]
@codex review

Please review the complete PR again at `9410f8dcfde2c75e93751be3f801b83e51e8e664` after the construction-counter validation fix. The previous finding is addressed in both parsers, with recomputed-checksum boundary tests through inspection, scalar reads and streaming. Two clean local review rounds, the incremental Release build and all 657 non-Ray checks passed. Ray/release revalidation is running in parallel. Start a fresh clean-review sequence on this commit; no compatibility modes or fallback paths are required.

**Comment by kaka11chen** [comment]
@codex review

Please perform the required second consecutive review of the complete PR on the same commit `9410f8dcfde2c75e93751be3f801b83e51e8e664`. The first clean pass is https://github.com/AstroVela/vane/pull/778#issuecomment-5569819150. All 678 affected tests passed on this commit; release rechecks retain only the three known #775 failures. Please review the full implementation again.

**Review by hubgeter — APPROVED** [review]

**Comment by kaka11chen** [comment]
@codex review

Please review the complete updated PR at commit `4e03545116b26019e73eb5e23f185ac91980ab28`, rebased onto upstream/main `04e68099e6` (the UInt8 IMAGE storage and ndarray materialization change). This is round 1 of the fresh review sequence on this head; the earlier clean reviews were for the pre-rebase commit. The conflict resolution retains exact frame/time/index behavior and adapts full-output assertions to the upstream Image contract.

**Comment by kaka11chen** [comment]
@codex review

Please perform the second complete review of unchanged commit `4e03545116b26019e73eb5e23f185ac91980ab28`. The first review of this rebased head completed without findings: https://github.com/AstroVela/vane/pull/778#issuecomment-5573336180 . Review the whole PR again for any remaining correctness or regression issues.

**Review by hubgeter — APPROVED** [review]
